### PR TITLE
feat: unref socket while idle

### PR DESCRIPTION
**Pull request by ronag**
Allow process shutdown once all requests have completed without
waiting for keepAliveTimeout.

### CODE REVIEW

**Comment by ronag** [comment]
Needs test.

**Comment by ronag** [comment]
@Ethan-Arrowood ping regarding your question on Slack.

**Review by mcollina — APPROVED** [review]
lgtm - having a test will be nice

**Comment by ronag** [comment]
Added test which unfortunately fails. @mcollina do you know how to debug/check what root ref keeps the process alive?

**Comment by ronag** [comment]
Maybe @jasnell has some tip on how to debug what ref is keep event loop alive?

**Comment by Ethan-Arrowood** [comment]
This looks good! I'll try and add some more test cases later today.

**Comment by mcollina** [comment]
have you tried https://github.com/mafintosh/why-is-node-running?

**Comment by ronag** [comment]
Yea, so the code works. The problem is the test. I think the server and client needs to run in a separate process for it to work. FYI @Ethan-Arrowood

File: test/client-unref.js [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+const { Client } = require('..')
+const http = require('http')
+
+test('end process on idle', (t) => {
+  t.plan(2)
+
+  const server = http.createServer((req, res) => {
+    res.end()
+  })
+  t.teardown(server.close.bind(server))
```
Comment (mcollina) [inline test/client-unref.js:13]: you need to remove this and unref() the server instead

**Review by mcollina — CHANGES_REQUESTED** [review]

File: test/client-unref.js (hunk 1 above)
Comment (ronag) [reply test/client-unref.js]: Tried that as well but it didn't help.

**Comment by Ethan-Arrowood** [comment]
I'm trying a multi-process test now using worker threads. will report back shortly with what I come up with

File: test/client-unref.js (hunk 1 above)
Comment (mcollina) [reply test/client-unref.js]: ☹️

**Comment by Ethan-Arrowood** [comment]
This is what I have so far: https://gist.github.com/Ethan-Arrowood/30341384a724f944a334b2cae7e4f739

Details in the gist title. Still hacking at it to see if I can get tap to play nicely.

**Comment by ronag** [comment]
I think you should put the server into the worker, not the other way around.

**Comment by Ethan-Arrowood** [comment]
worker_thread test PR is up against this branch (from my fork)

**Comment by Ethan-Arrowood** [comment]
irregardless of the worker thread test. The existing connect-timeout test is failing with this new change. Any ideas? Do we need to keep the client referenced while the `request` call fails?

File: test/client-unref.js [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+    body.resume()
+    // once the body has been read entirely set a timeout cb that will fail the
+    // tap test if it executes. It will execute if the client does not
+    // unreference itself as well.
+    await once(body, 'end')
+    setTimeout(() => {
+      t.fail('Client did not automatically unref')
+    }, 2e3).unref()
+    // clean up the server in the worker thread so the tap test can end
+    worker.postMessage('close_server')
```
Comment (ronag) [inline test/client-unref.js:33]: The problem here is that when the server closes the socket will die, thus we are not actually testing the unref case here...? Maybe I was wrong and the client should be in the worker?

Comment (Ethan-Arrowood) [reply test/client-unref.js:33]: good point. i'll revisit in a bit and try again.

**Comment by ronag** [comment]
@Ethan-Arrowood I think I managed to fix it.

**Review by mcollina — APPROVED** [review]
lgtm

**Comment by Ethan-Arrowood** [comment]
Which release will this be published in? 3 or 4?

**Comment by ronag** [comment]
4

**Pull request merged**